File: app/src/main/java/com/example/alertdrive/Fragment/FragmentBottom.java
package com.example.alertdrive.Fragment;

import android.app.FragmentTransaction;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.TextView;

import androidx.annotation.Nullable;
import android.app.Fragment;

import com.example.alertdrive.Custom_Adapter;
import com.example.alertdrive.GetTodoList;
import com.example.alertdrive.MainActivity;
import com.example.alertdrive.R;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ExecutionException;

public class FragmentBottom extends Fragment {


    MainActivity context;

    List<String> todo = new ArrayList<>();
    List<Boolean> compl = new ArrayList<>();
    JSONArray todos;
    List<JSONObject> alltodos = new ArrayList<JSONObject>();
    List<JSONObject> dynTodos = new ArrayList<JSONObject>();
    List<JSONObject> compltodos = new ArrayList<JSONObject>();
    List<JSONObject> incompltodos = new ArrayList<JSONObject>();
    Set<JSONObject> uniqueEntries = new HashSet<JSONObject>();
    Custom_Adapter exAdapter2;
    // This method will be invoked when the Fragment view object is created.
    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        context = (MainActivity) getActivity();
        String strtext = "";
        String title = "";

        View retView = inflater.inflate(R.layout.fragment_bottom, container);
        ListView listView =  retView.findViewById(android.R.id.list);
        List<String> todo = new ArrayList<>();
        List<Boolean> compl = new ArrayList<>();


        GetTodoList asyncTask=new GetTodoList();
        try {
            strtext = asyncTask.execute().get();


            // Getting JSON Array node
            todos = new JSONArray(strtext);
            int temp=0;
            int temp1=0;
            for (int i = 0; i < todos.length(); ++i) {
                JSONObject c = todos.getJSONObject(i);
                alltodos.add(c);
                boolean id = c.getBoolean("completed");
                if (id) {
                    compltodos.add(c);
                }else {
                    incompltodos.add(c);
                }
            }
            dynTodos = alltodos;

            // looping through All Contacts
            for (int i = 0; i < 5; i++) {
                JSONObject c = dynTodos.get(i);
                todo.add(c.getString("title"));
                compl.add(c.getBoolean("completed"));


            }
            exAdapter2 = new Custom_Adapter(todo, getContext(), compl);

            listView.setAdapter(exAdapter2);
            if(retView!=null) {
            }

        } catch (ExecutionException e) {
            e.printStackTrace();
        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (JSONException e) {
            e.printStackTrace();
        }

        if(retView!=null) {
        }
        return retView;
    }

    public void update(int page){
        try{
            todo.clear();
            compl.clear();
            for (int i = page; i < page+5; i++) {
                JSONObject c = dynTodos.get(i);
                todo.add(c.getString("title"));
                compl.add(c.getBoolean("completed"));
            }
            exAdapter2.update(todo,compl);
            } catch (JSONException e) {
                e.printStackTrace();
            }
    }

    public void filterall(){
        try{
            dynTodos = alltodos;
            todo.clear();
            compl.clear();
            for (int i = 0; i < 5; i++) {
                JSONObject c = dynTodos.get(i);
                todo.add(c.getString("title"));
                compl.add(c.getBoolean("completed"));
            }
            exAdapter2.update(todo,compl);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }
    public void filtercompleted(){
        try{
            dynTodos = compltodos;
            todo.clear();
            compl.clear();
            for (int i = 0; i < 5; ++i) {
                JSONObject c = dynTodos.get(i);
                todo.add(c.getString("title"));
                compl.add(c.getBoolean("completed"));
            }
            exAdapter2.update(todo,compl);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void filterIncompleted(){
        try{
            dynTodos = incompltodos;
            todo.clear();
            compl.clear();
            for (int i = 0; i < 5; ++i) {
                JSONObject c = dynTodos.get(i);
                todo.add(c.getString("title"));
                compl.add(c.getBoolean("completed"));
            }
            exAdapter2.update(todo,compl);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void filtersearch(String text){
        String[] words = text.split(" ");
        try{

            for (int i = 0; i < alltodos.size(); ++i) {
                JSONObject c = alltodos.get(i);
                String sentence = c.getString("title");
                for(String word : words){
                    if (sentence.contains(word)) {
                        uniqueEntries.add(c);
                    }
                }
            }
            dynTodos.clear();
            for(JSONObject i : uniqueEntries) {
                dynTodos.add(i);
            }

            todo.clear();
            compl.clear();
            for (int i = 0; i < 5; ++i) {
                JSONObject c = dynTodos.get(i);
                todo.add(c.getString("title"));
                compl.add(c.getBoolean("completed"));
            }
            exAdapter2.update(todo,compl);
        } catch (JSONException e) {
            e.printStackTrace();
        }

    }



}

File: app/src/main/java/com/example/alertdrive/Fragment/FragmentBottom.kt
package com.example.alertdrive.Fragment

import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ListView
import android.app.Fragment
import android.widget.Toast

import com.example.alertdrive.Custom_Adapter
import com.example.alertdrive.GetTodoList
import com.example.alertdrive.MainActivity
import com.example.alertdrive.R
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.ArrayList
import java.util.HashSet
import java.util.concurrent.ExecutionException

class FragmentBottom : Fragment() {



    internal var todo: MutableList<String> = ArrayList()
    internal var compl: MutableList<Boolean> = ArrayList()
    internal  var todos: JSONArray? = null
    internal var alltodos: MutableList<JSONObject> = ArrayList()
    internal var dynTodos: MutableList<JSONObject> = ArrayList()
    internal var compltodos: MutableList<JSONObject> = ArrayList()
    internal var incompltodos: MutableList<JSONObject> = ArrayList()
    internal var uniqueEntries: MutableSet<JSONObject> = HashSet()
    internal  var exAdapter2: Custom_Adapter? = null
    internal var asyncTask = GetTodoList()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val context = activity as MainActivity
        var strtext = ""

        val retView = inflater.inflate(R.layout.fragment_bottom, container)
        val listView = retView!!.findViewById<ListView>(android.R.id.list)
        val todo = ArrayList<String>()
        val compl = ArrayList<Boolean>()
        val asyncTask = GetTodoList()

        try {
            strtext = asyncTask.execute().get()
            todos = JSONArray(strtext)
            for (i in 0 until todos!!.length()) {
                val c = todos!!.getJSONObject(i)
                alltodos.add(c)
                val id = c.getBoolean("completed")
                if (id) {
                    compltodos.add(c)
                } else {
                    incompltodos.add(c)
                }
            }
            dynTodos = alltodos
            for (i in 0..4) {
                val c = dynTodos[i]
                todo.add(c.getString("title"))
                compl.add(c.getBoolean("completed"))


            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                exAdapter2 = Custom_Adapter(todo, getContext(), compl)
            }

            listView.adapter = exAdapter2
            if (retView != null) {
            }

        } catch (e: ExecutionException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        if (retView != null) {
        }
        return retView
    }

    fun update(page: Int) {
        try {
            todo.clear()
            compl.clear()
            if (page + 5 < dynTodos.size) {
                for (i in page until page + 5) {
                    val c = dynTodos[i]
                    todo.add(c.getString("title"))
                    compl.add(c.getBoolean("completed"))
                }
                exAdapter2!!.update(todo, compl)
            } else {
                Toast.makeText(activity, "REACHED END", Toast.LENGTH_SHORT).show()
            }
        } catch (e: JSONException) {
            e.printStackTrace()
        }

    }

    fun filterall() {
        retrieveDaTA()
        updateFunctionality(alltodos)
    }

    fun filtercompleted() {
        retrieveDaTA()
        updateFunctionality(compltodos)
    }

    fun filterIncompleted() {
        retrieveDaTA()
        updateFunctionality(incompltodos)
    }

    fun filtersearch(text: String) {
        val words = text.split(" ".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
        try {
            uniqueEntries.clear()
            for (i in alltodos.indices) {
                val c = alltodos[i]
                val sentence = c.getString("title")
                for (word in words) {
                    if (sentence.contains(word)) {
                        uniqueEntries.add(c)
                    }
                }
            }
            dynTodos.clear()
            if (uniqueEntries.size == 0) {
                Toast.makeText(activity, "TODO NOT FOUND", Toast.LENGTH_SHORT).show()
            } else {
                for (i in uniqueEntries) {
                    dynTodos.add(i)
                }
            }


            updateFunctionality(dynTodos)
        } catch (e: JSONException) {
            e.printStackTrace()
        }

    }

    fun updateFunctionality(`val`: List<JSONObject>) {
        try {
            dynTodos = `val` as MutableList<JSONObject>
            todo.clear()
            compl.clear()
            if (dynTodos.size > 0) {
                for (i in 0..4) {
                    val c = dynTodos[i]
                    todo.add(c.getString("title"))
                    compl.add(c.getBoolean("completed"))
                }
            } else {
                retrieveDaTA()
                updateFunctionality(alltodos)
            }
            exAdapter2?.update(todo, compl)
        } catch (e: JSONException) {
            e.printStackTrace()
        }

    }

    internal fun retrieveDaTA() {
        try {
            val strtext = GetTodoList().execute().get()
            alltodos.clear()
            compltodos.clear()
            incompltodos.clear()
            todos = JSONArray(strtext)
            for (i in 0 until todos!!.length()) {
                val c = todos!!.getJSONObject(i)
                alltodos.add(c)
                val id = c.getBoolean("completed")
                if (id) {
                    compltodos.add(c)
                } else {
                    incompltodos.add(c)
                }
            }
            dynTodos = alltodos
        } catch (e: ExecutionException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: JSONException) {
            e.printStackTrace()
        }

    }


}
